File: packages/java/src/main/java/com/missioninbox/resources/EmailQueue.java
package com.missioninbox.resources;

import com.missioninbox.http.HttpClient;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * The {@code emailQueue} resource. Access via {@code mi.emailQueue}.
 */
public final class EmailQueue {

    private final HttpClient http;

    public EmailQueue(HttpClient http) {
        this.http = http;
    }

    /**
     * List queued messages.
     *
     * <p>Recognised keys: {@code status}, {@code sendingAccountId}, {@code page}, {@code limit}.
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> list(Map<String, Object> params) {
        return (Map<String, Object>) http.request("GET", "/api/email/queue", params, null);
    }

    /** Retry a failed queued message. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> retry(String id) {
        return (Map<String, Object>) http.request(
                "POST", "/api/email/queue/" + encode(id) + "/retry", null, null);
    }

    /** Cancel a pending queued message. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> cancel(String id) {
        return (Map<String, Object>) http.request(
                "POST", "/api/email/queue/" + encode(id) + "/cancel", null, null);
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }
}

File: packages/python/src/missioninbox/resources/email_queue.py
"""The `email_queue` resource. Access via `mi.email_queue`."""

from __future__ import annotations

from typing import Any, Optional
from urllib.parse import quote

from .._http import HttpClient


class EmailQueue:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        status: Optional[str] = None,
        sending_account_id: Optional[str] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> dict[str, Any]:
        """List queued messages."""
        query: dict[str, Any] = {}
        if status is not None:
            query["status"] = status
        if sending_account_id is not None:
            query["sendingAccountId"] = sending_account_id
        if page is not None:
            query["page"] = page
        if limit is not None:
            query["limit"] = limit
        return self._http.request("GET", "/api/email/queue", query=query)

    def retry(self, id: str) -> dict[str, Any]:
        """Retry a failed queued message."""
        return self._http.request("POST", f"/api/email/queue/{quote(id, safe='')}/retry")

    def cancel(self, id: str) -> dict[str, Any]:
        """Cancel a pending queued message."""
        return self._http.request("POST", f"/api/email/queue/{quote(id, safe='')}/cancel")
